Бочаров Александр

Geratoptus opened the pull request.

Geratoptus committed:
Без лишних слов - мужчина умер...

Geratoptus committed:
Доливка unversioned files

Folleach commented on cs/Markdown/Tokenizer/Scanners/NumberScanner.cs:
```diff
@@ -0,0 +1,17 @@
+{
+    public Token? Scan(string text, int begin = 0)
+    {
+        var numberEnumerable = text
+            .Skip(begin)
+            .TakeWhile(CanScan);
```
На сколько я понял, в `text` всегда заходит оригинальный текст.  
Удовлетворяет ли `Skip` сложности `O(n)` в таком случае?

Красивее было реализовать `ITokenScanner` с использованием ссылки на кусок текста. В C# для этого существуюет [Span\<T\>](https://learn.microsoft.com/en-us/dotnet/standard/memory-and-spans/)

Сделаешь?
Тогда не понадобится сохранять контекст дополнительными параметрами вроде `begin`

Folleach replied:
resolved

Folleach commented on cs/MarkdownTests/MdTest.cs:
```diff
@@ -0,0 +1,49 @@
+        
+        Enumerable.Range(1, results.Count - 1)
+            .Select(i => results[i].Ticks / results[i - 1].Ticks)
+            .Should().OnlyContain(timeRatio => timeRatio < scale * scale);
```
Может будем сравнивать с меньшей асимптотичной сложностью?
`O(n^2)` выглядит много

Folleach replied:
resolved

Folleach replied:
resolved

Geratoptus committed:
fix(MdTests.cs): пофиксил тест на скорость работы:
- Убрал ненужные классы
- В сравнении добавил каст к (double), теперь проверка корректная
- Добавил GC.Collect() перед замером времени для более корректной проверки

Folleach commented on cs/MarkdownTests/Parser/Rules/ItalicRuleTest.cs:
```diff
@@ -0,0 +1,114 @@
+        node.Should().NotBeNull();
+        node.Children.Should().ContainSingle(n => n.NodeType == NodeType.Text);
+        return node.ToText(tokens);
+    }
```
Было бы здорово проверить, что сама нода `Italic`

Folleach replied:
resolved

Folleach commented on cs/MarkdownTests/MdAcceptanceTest.cs:
```diff
@@ -0,0 +1,37 @@
+
+namespace MarkdownTests;
+
+public class MdAcceptanceTest
```
Добавлю тут странные примеры, которые у меня получилось найти:

1. `\\_emmm_` превращается в `\_emmm_`.
   Вроде бы символ `\` экранировали, поэтому `emmm` должен был обернуться в html тег `em`
2. `_e\mmm_` превращается в `_e\mmm_`, хотя экранирование не касается тегов, почему они не преобразовались?
3. В задании не говорилось, но абзацы могут быть не только \<h1>
   # h1
   ## h2
   ### h3
   #### h4

### Задачи со звёздочкой
1. ​
   ![image](https://github.com/user-attachments/assets/491d5a92-9b4e-4147-8e91-ea28801cd282)
2. ​
   ![image](https://github.com/user-attachments/assets/5eb48f89-197b-4054-9d84-29278c74e6b0)

Folleach replied:
resolved

Folleach commented on cs/Markdown/Parser/Rules/InWordBoldRule.cs:
```diff
@@ -0,0 +1,48 @@
+            new KleeneStarRule(valueRule),
+            PatternRuleFactory.DoubleUnderscore(),
+        ]);
+        var continuesRule = new OrRule(possibleContinues);
```
А почему эти правила создаются каждый вызов метода `Match`?  
Вроде бы они от данных не зависят, и могут быть созданы один раз

Folleach replied:
resolved

Geratoptus committed:
feat(Scanners): переписал сканеры на Memory<char>, теперь работают быстрее

Geratoptus committed:
fix(ScannerTests): поправил тесты для новой реализации сканеров

Geratoptus committed:
feat(ListOrderExtensions): написал свой скип, который прыгает сразу куда надо, а не как в LINQ

Geratoptus committed:
feat(PatternRule|TextRule): теперь используется быстрый Skip, квадрата больше нет

Geratoptus committed:
feat(MdTokenizer): внедрил новые сканеры в токенизатор

Geratoptus committed:
feat(MdTests.cs): увеличил размеры генерируемых Markdown-разметок в тесте на скорость

Geratoptus committed:
feat(ItalicRuleTest/ShouldMatch_TagInWord): добавил проверку на тип ноды

Geratoptus committed:
fix(SpecScanner): теперь правильно сканит неразрывный пробел и пробел нулевой ширины

Geratoptus committed:
fix(ParagraphRule): теперь корректно матчит кейсы с эскейпнутым бэкслэшем

Geratoptus committed:
fix(BoldRule.cs): теперь правило корректно мэтчит тэг с внутренним слэшем

Geratoptus committed:
fix(ItalicRule|InWordItalicRule): теперь правило корректно мэтчит тэг с внутренним слэшем

Geratoptus committed:
refactor(Rules): все что можно проинициализировать без дополнительных переданных параметров вынес в статические поля

Geratoptus committed:
feat(HeaderRule.cs): теперь правило мэтчить все 6 хедеров

Geratoptus committed:
feat(HtmlGenerator.cs): теперь HtmlGenerator может генерировать все 6 html-заголовков

Geratoptus committed:
fix(ItalicRule|BoldRule): теперь правила корректно мэтчат, когда внутри тэга есть * и #

Geratoptus committed:
fix(BodyRuleTest.cs): пофиксил тест с новым хедером

Geratoptus committed:
feat(HeaderRuleTest.cs): добавил тестов на мэтчинг новых заголовков

Geratoptus committed:
feat(ItalicRuleTest.cs): добавил тестов для италика на спец символы

Geratoptus committed:
feat(BoldRuleTest.cs): нашел тесты, на которых не работает

Geratoptus committed:
fix(InWorldBoldRule): теперь правило корректно мэтчит спец символы, когда тэг внутри слова

Geratoptus committed:
fix(MdTest.cs): теперь асимптотическая сложность сравнивается с N*Log2(N), а не с N*N

Geratoptus committed:
feat(MdAcceptanceTest): накидал случаев под новый функционал и пофикшенные проблемы

Geratoptus committed:
refactor(InWordBoldRule): удалил забытый коммент

Geratoptus committed:
refactor(MdTokenizerTests): поправил неймспейсы

Folleach commented on cs/Markdown/Tokenizer/MdTokenizer.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        var begin = 0;
+        var tokenList = new List<Token>();
+        var memoryText = new Memory<char>(text.ToCharArray());
```
https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.asmemory?view=net-9.0

Чтобы избежать аллокации дополнительных O(text) памяти
